Buttonclass

jay4kelly opened the pull request:
Closes #838 

#### What has been done to verify that this works as intended?
Tested using UI for all buttons and chevrons. 
#### Why is this the best possible solution? Were any other approaches considered?
Created classes called DetailsStatusButton and ExportConfigurationButton. The selection column didn't need a class for it as there was no business logic incorporated. So we continued to use a compareSelectionfunction inside the UI class.
#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
Localized just to the UI tests conducted.
### Does this change require updates to documentation? If so, please file an issue at https://github.com/opendatakit/docs/issues/new and include the link below.
No

ggalmazor commented on src/org/opendatakit/briefcase/ui/reused/ExportConfigurationButton.java:
```diff
@@ -0,0 +1,62 @@
+/**
+ * Contains application logic pertaining to the Export Configuration Button
+ */
+public class ExportConfigurationButton extends JButton implements Comparable<ExportConfigurationButton> {
```
The following may seem strange, but we always want to stay away from extending classes when we're not implementing Polymorphism, which is not the case in this instance, because we're not really trying to [replace conditional logic](https://refactoring.com/catalog/replaceConditionalWithPolymorphism.html).

I've gone ahead and prepared an alternative implementation of `ExportConfigurationButton` that wraps around a `Jbutton` instance and has the same behavior without extending `JButton`.  You can check it here: https://github.com/ggalmazor/briefcase/commit/e7bb0891403a54a13892f9d6439f753e9536d363

jay4kelly commented:
Changed DetailStatusButton and ExportConfigurationButton to not extend JButton. Added ButtonProcessing Interface. Could have added a CommonButton class for the button classes to extend instead of using the interface. I think this would require us to remove the static create method on the button classes and use a new operator instead.

ggalmazor commented on src/org/opendatakit/briefcase/ui/export/components/ExportFormsTableViewModel.java:
```diff
@@ -115,14 +104,15 @@ private void removeConfiguration(FormStatus form) {
-  private void updateConfButton(FormStatus form, JButton button) {
-    button.setForeground(forms.hasConfiguration(form) ? NO_CONF_OVERRIDE_COLOR : DARK_GRAY);
+  private void updateConfButton(FormStatus form, ExportConfigurationButton button) {
+    button.setStatus(forms.hasConfiguration(form));
```
I think `button.setStatus()` is kinda unexpected because the meaning of "status" in a conf button is different than the one in a details button. I believe you had to have the same name for both classes because they extend it from `ButtonProcessing`, but this is not really a requirement, since you're using the subclasses (`DetailStatusButton` and `ExportConfigurationButton`) instead of the interface, thus not doing Polymorphism (which is totally fine here, by the way).

This is telling us that the `setStatus` method is not part of the `ButtonProcessing` interface. I'd suggest you remove it from the interface and take the chance to give it a more specific name in `ExportConfigurationButton` e.g. `setConfigured`.

jay4kelly replied:
Sure.

ggalmazor commented on src/org/opendatakit/briefcase/ui/export/components/ExportFormsTableViewModel.java:
```diff
@@ -157,7 +147,7 @@ public Object getValueAt(int rowIndex, int columnIndex) {
 
   @Override
-  // Suppressing next ParameterName checkstyle error becasue 'aValue' param triggers it by mistake
+  // Suppressing next ParameterName checkstyle error because 'aValue' param triggers it by mistake
```
Good catch! ;)

ggalmazor commented on src/org/opendatakit/briefcase/ui/reused/ButtonProcessing.java:
```diff
@@ -0,0 +1,33 @@
+ */
+public interface ButtonProcessing {
+
+  void onClick(Runnable callback);
```
I believe that `getJButton()` would be the only method that needs to exist in this interface. All the other methods are accessed using the subclasses.

Another method in this interface could be the one I suggested in a previous comment i.e. `void click(Runnable callback)` method.

jay4kelly replied:
Still have the onClick method in the interface.  DId not fully understand how to implement the change of button.addActionListener(__ ->   to  button.onClick(().
Can you explain how the "__" reference is used? I have not seen that before.

ggalmazor replied:
My bad! This is something I do when I'm forced to declare an arg that I will ignore in the lambda's body block. `__` is simply the shortest name a variable can have using `_` characters, and the goal would be to implicitly say that the can be ignored.

ggalmazor commented on src/org/opendatakit/briefcase/ui/reused/DetailsStatusButton.java:
```diff
@@ -0,0 +1,72 @@
+import java.awt.Insets;
+import javax.swing.JButton;
+
+public class DetailsStatusButton implements ButtonProcessing, Comparable<ButtonProcessing> {
```
I believe implementing `Comparable<ButtonProcessing>` is not needed because we will use a reference to a static method.

jay4kelly replied:
TableRowSorter is referencing the column class of the table column.  And as long as that class implements the Comparable interface, there is no need to pass in a comparator.

ggalmazor replied:
Huh! I didn't notice that... Thanks for explaining :)

ggalmazor commented on src/org/opendatakit/briefcase/ui/reused/DetailsStatusButton.java:
```diff
@@ -0,0 +1,72 @@
+    button.setForeground(status ?  DARK_GRAY : LIGHT_GRAY );
+  }
+
+  public int compareTo(ButtonProcessing b) {
```
So, this method should be migrated to `public static int compare(DetailsStatusButton a, DetailsStatusButton b)`

jay4kelly replied:
See earlier comment on Comparable

ggalmazor commented on src/org/opendatakit/briefcase/ui/reused/ExportConfigurationButton.java:
```diff
@@ -0,0 +1,74 @@
+/**
+ * Contains application logic pertaining to the Export Configuration Button
+ */
+public class ExportConfigurationButton implements ButtonProcessing, Comparable<ButtonProcessing> {
```
Same, here: we shouldn't need to implement `Comparable`, and the `compareTo` method should be static and take two `ExportConfigurationButton` objects

ggalmazor commented on src/org/opendatakit/briefcase/ui/reused/UI.java:
```diff
@@ -62,24 +59,18 @@
 
-  @SuppressWarnings("checkstyle:AvoidEscapedUnicodeCharacters")
-  public static JButton buildDetailButton(FormStatus form) {
+  public static DetailsStatusButton buildDetailButton(FormStatus form) {
```
I would expect this to be a static factory in `DetailsStatusButton`

jay4kelly replied:
Agreed. Made the change.

ggalmazor commented on src/org/opendatakit/briefcase/ui/reused/transfer/TransferFormsTableView.java:
```diff
@@ -72,6 +74,7 @@
     setFillsViewportHeight(true);
 
     TableRowSorter<TransferFormsTableViewModel> sorter = sortBy(getModel(), FORM_NAME_COL, ASCENDING);
+    sorter.setComparator(SELECTED_CHECKBOX_COL, (Comparator<Boolean>) UI::compareSelectionButton);
```
I believe we need to set another comparator for the details column as in:
```
sorter.setComparator(DETAIL_BUTTON_COL, (Comparator<DetailsStatusButton>) DetailsStatusButton::compare);

jay4kelly replied:
See earlier comment on Comparable

ggalmazor commented on src/org/opendatakit/briefcase/ui/reused/transfer/TransferFormsTableView.java:
```diff
@@ -86,8 +89,8 @@ private void relayClickToButton(MouseEvent event) {
-      if (value instanceof JButton)
-        ((JButton) value).doClick();
+      if (value instanceof ButtonProcessing)
+        ((ButtonProcessing) value).getJButton().doClick();
```
I'd expect `((ButtonProcessing) value).click();` here.

ggalmazor commented on src/org/opendatakit/briefcase/ui/export/components/ExportFormsTableView.java:
```diff
@@ -80,6 +83,7 @@
     setFillsViewportHeight(true);
 
     TableRowSorter<ExportFormsTableViewModel> sorter = sortBy(getModel(), FORM_NAME_COL, ASCENDING);
+    sorter.setComparator(SELECTED_CHECKBOX_COL, (Comparator<Boolean>) UI::compareSelectionButton);
```
I believe we need to set a couple more comparators for the details and conf button columns as in:
```
sorter.setComparator(OVERRIDE_CONF_COL, (Comparator<ExportConfigurationButton>) ExportConfigurationButton::compare);
sorter.setComparator(DETAIL_BUTTON_COL, (Comparator<DetailsStatusButton>) DetailsStatusButton::compare);

jay4kelly replied:
See earlier comment on Comparable

jay4kelly commented on src/org/opendatakit/briefcase/ui/export/components/ExportFormsTableView.java:
```diff
@@ -90,8 +94,8 @@ private void relayClickToButton(MouseEvent event) {
-      if (value instanceof JButton)
-        ((JButton) value).doClick();
+      if (value instanceof ButtonProcessing)
+        ((ButtonProcessing) value).getJButton().doClick();
```
Tried that but the doClick call was required to be performed by a JButton class. We can recheck that if you think it should work.

ggalmazor replied:
I was thinking about adding a `ButtonProcessing.doClick` method that wraps `button.doClick`, to avoid breaking encapsulation.

ggalmazor commented:
Hi, @jay4kelly!  

After exploring a bit more the changes I was suggesting, I've come to the conclusion that they don't bring that much value overall.

This PR is looking good to me now :) Thanks for your work!!

jay4kelly commented:
Closes #838

kkrawczyk123 commented:
**Tested with success!**
I haven't' noticed any regression. I have confirmed that the issue identified in #830 has been fixed. 
Verified on Ubuntu, MacOS and Windows.

@opendatakit-bot unlabel "needs testing"
@opendatakit-bot label "behavior verified"

ggalmazor commented:
Hi, @jay4kelly! I've rebased your branch on top of master to be able to merge it. Thanks for your work!
